### PR TITLE
Refactor Digest

### LINKED ISSUE #9483: Invalid Downcast Base+ <- Extended in non-captured block with restrictions
Code like the following worked prior to 0.35:
```crystal
require "digest/sha1"
 
Digest::SHA1.hexdigest do |digest|
  while rand > 0.5 # obviously, in production this is a more useful condition
    digest.update("")
  end
end
```

The error I get is:
```
$ crystal run test.cr
BUG: trying to downcast Digest::Base+ <- Digest::SHA1 (Exception)
  from raise<Exception>:NoReturn
  from raise<String>:NoReturn
  from Crystal::CodeGenVisitor#downcast_distinct<LLVM::Value, Crystal::Type+, Crystal::Type+>:NoReturn
  from Crystal::CodeGenVisitor#downcast<LLVM::Value, Crystal::Type+, Crystal::Type+, Bool>:LLVM::Value
  from Crystal::CodeGenVisitor#visit<Crystal::Var+>:(LLVM::Value | Nil)
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::CodeGenVisitor#prepare_call_args_non_external<Crystal::Call, Crystal::Def+, Crystal::Type+>:Tuple(Array(LLVM::Value), Bool)
  from Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::CodeGenVisitor#visit<Crystal::Assign>:(Bool | Nil)
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::CodeGenVisitor#visit<Crystal::Call>:Bool
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::ASTNode+@Crystal::ASTNode#accept<Crystal::CodeGenVisitor>:Nil
  from Crystal::Compiler#codegen<Crystal::Program, Crystal::ASTNode+, Array(Crystal::Compiler::Source), String>:(Tuple(Array(Crystal::Compiler::CompilationUnit), Array(String)) | Nil)
  from Crystal::Compiler#compile<Array(Crystal::Compiler::Source), String>:Crystal::Compiler::Result
  from Crystal::Command#run_command<Bool>:Nil
  from Crystal::Command#run:(Bool | Crystal::Compiler::Result | Nil)
  from __crystal_main
  from main
Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/crystal-lang/crystal/issues
```

Link to playground reproducing the error: https://play.crystal-lang.org/#/r/99si

I am testing with 0.35 on OSX 10.14.6 (it works on 0.34 and earlier).

### LINKED ISSUE #9699: OpenSSL::DigestIO.hexstring raising deprecation warning
Side effect of #8426. Invoking `hexstring` on `OpenSSL::DigestIO` raising deprecation warning
```
In /usr/local/Cellar/crystal/0.35.1_1/src/openssl/digest/digest_io.cr:23:5

 23 | delegate digest, hexdigest, base64digest, to: @digest_algorithm
      ^
Warning: expanding macro


There was a problem expanding macro 'delegate'

Called macro defined in /usr/local/Cellar/crystal/0.35.1_1/src/object.cr:1245:3

 1245 | macro delegate(*methods, to object)

Which expanded to:

    1 |
    2 |
    3 |         def digest(*args, **options)
    4 |           @digest_algorithm.digest(*args, **options)
    5 |         end
    6 |
    7 |
    8 |           def digest(*args, **options)
    9 |             @digest_algorithm.digest(*args, **options) do |*yield_args|
   10 |               yield *yield_args
   11 |             end
   12 |           end
   13 |
   14 |
   15 |
   16 |
   17 |         def hexdigest(*args, **options)
 > 18 |           @digest_algorithm.hexdigest(*args, **options)
   19 |         end
   20 |
   21 |
   22 |           def hexdigest(*args, **options)
   23 |             @digest_algorithm.hexdigest(*args, **options) do |*yield_args|
   24 |               yield *yield_args
   25 |             end
   26 |           end
   27 |
   28 |
   29 |
   30 |
   31 |         def base64digest(*args, **options)
   32 |           @digest_algorithm.base64digest(*args, **options)
   33 |         end
   34 |
   35 |
   36 |           def base64digest(*args, **options)
   37 |             @digest_algorithm.base64digest(*args, **options) do |*yield_args|
   38 |               yield *yield_args
   39 |             end
   40 |           end
   41 |
   42 |
   43 |
   44 |
Warning: Deprecated OpenSSL::Digest#hexdigest. Use `final.hexstring` instead.

In /usr/local/Cellar/crystal/0.35.1_1/src/digest/base.cr:143:5

 143 | digest.hexstring
       ^-----
Warning: Deprecated OpenSSL::Digest#digest. Use `final` instead.

A total of 2 warnings were found.
```

# Sample code
```Crystal
require "openssl"

io = IO::Memory.new("Hello Crystal")
dio = OpenSSL::DigestIO.new(io, "SHA256")
dio.gets_to_end
pp dio.hexdigest
```
# Crystal Version
```
Crystal 0.35.1 (2020-06-19)

LLVM: 10.0.0
Default target: x86_64-apple-macosx
```

### PR DESCRIPTION
**Pull request by bcardiff**
This PR takes implements the left ideas to polish the Digest module and hide a bit OpenSSL when dealing with digests.

* `Digest::{MD5, SHA1, SHA256, SHA512}` are all backed by the `OpenSSL::Digest`.
* `OpenSSL::Digest` is still available if you need to build other digest as MD4, RIPEMD160, SHA224, SHA384.
* The former pure Crystal MD5, SHA1 are moved to `Crystal::Digest::{MD5, SHA1}` and should be treated as internal compiler implementation.
* Some convenient methods to update a digest from IO / Files existed in `OpenSSL::Digest::Base`. These are integrated for all digests and that class is removed.
* `Digest` is now an abstract base class, instead of `Digest::Base`.
* `Digest::ClassMethods` module is introduced to group all digests classes that have argless `.new` method (`OpenSSL::Digest` is not one of these and so it can not have class methods as `Digest::MD5.hexdigest("foo")`).
* The former `OpenSSL::DigestIO` is refactored to `IO::Digest` and can be used with any `Digest`. Note: Although the name is `IO::Digest`, it available when `require "digest"` or `require "digest/io"`. `require "io"` is part of the prelude, but digest is not.

Closes #9483
Closes #9699
Fixes #9697
Ref https://github.com/crystal-lang/crystal/pull/9496#issuecomment-645998949

### CODE REVIEW

**Comment by asterite** [comment]
Can we consider renaming `Digest::Algorithm` to `Digest::ClassMethods`? I just find the name "algorithm" to be arbitrary and not at all related to "these are the class methods for a digest".

File: src/digest/digest.cr [hunk 1]
```diff
@@ -1,12 +1,18 @@
 require "base64"
 
-class Digest::FinalizedError < Exception
-end
+# The Digest module contains implementations of hashing algorithms like
```
Comment (straight-shoota) [inline src/digest/digest.cr:3]: This doc comment could be more elaborate and refer to `Digest` as the base type for implementations (instead of an container).

**Review by straight-shoota — COMMENTED** [review]
This looks really great ❤️ 

My only concern is the second constructor of `IO::Digest` which receives the algorithm name as a string. This seems odd considering this doesn't work anywhere else. If we want string-based algorithm resolution, it should probably be a generic implemention on `Digest`. A general purpose implementation should also be open for algorithms not backed by open ssl (so it would probably need some kind of plugin registry?).
The bottom line is: This is not an essential feature and needs more thinking through. So I'd rather leave it out of this PR.

**Comment by bcardiff** [comment]
Suggestions addressed.

And I went through and drop the already deprecated method so we are in a clean state for 1.0 in Digest.

File: src/digest/md5.cr [hunk 2]
```diff
@@ -1,254 +1,24 @@
...
-      dst[ii + 3] = ((@buf[i] >> 24) & 0xFF).to_u8
-      ii += 4
-    end
-  end
-
-  def digest_size : Int32
-    16
+  def dup
+    self.class.new(dup_ctx)
   end
```
Comment (Sija) [inline src/digest/md5.cr:23]: Couldn't this be defined directly within the `Digest` class itself?

File: src/http/web_socket/protocol.cr [hunk 3]
```diff
@@ -326,7 +328,7 @@ class HTTP::WebSocket::Protocol
 
   def self.key_challenge(key)
     {% if flag?(:without_openssl) %}
-      Digest::SHA1.base64digest(key + GUID)
+      ::Crystal::Digest::SHA1.base64digest(key + GUID)
```
Comment (Sija) [inline src/http/web_socket/protocol.cr:331]: IIUC this won't work without `: without_openssl` flag.

File: src/digest/md5.cr (hunk 2 above)
Comment (bcardiff) [reply src/digest/md5.cr:23]: No, becuase the `OpenSSL::Digest` has no constructor without algorithm name. And having another intermediate class sounds overkill to me.

File: src/http/web_socket/protocol.cr (hunk 3 above)
Comment (bcardiff) [reply src/http/web_socket/protocol.cr:331]: Exactly. `-Dwithout_openssl` will use the 100% crystal sha1 implementation. But by default the openssl one is used. The only thing that is changing here is the location of the crystal sha1 implementation.

**Review by straight-shoota — APPROVED** [review]

File: src/http/web_socket/protocol.cr (hunk 3 above)
Comment (Sija) [reply src/http/web_socket/protocol.cr:331]: How come? Does `::Crystal::Digest::SHA1` delegates to `OpenSSL::...` somehow?

Comment (bcardiff) [reply src/http/web_socket/protocol.cr:331]: `::Crystal::Digest::SHA1` is the pure crystal implementation (formerly `::Digest::SHA1`).

After this PR `::Digest::SHA1` uses openssl, but `::Crystal::Digest::SHA1` does not.

Comment (Sija) [reply src/http/web_socket/protocol.cr:331]: That's what I'm asking about since the `require` is conditional.

Comment (straight-shoota) [reply src/http/web_socket/protocol.cr:331]: So is the instantiation.

Comment (Sija) [reply src/http/web_socket/protocol.cr:331]: @straight-shoota So true, somehow I missed that 🤦‍♂️
